### ISSUE TITLE
[Blazor WASM] Upgrading project to .NET 8 and breaks [Authorize] Attribute

### ISSUE THREAD

**Issue by patrickpham3339** [dotnet/aspnetcore#53075]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Whom it may concerned,

My project is upgrading to Net 8 and using `@attribute [Authorize]`. when I logged out and called NotifyAuthenticationStateChanged method by using CustomAuthenticationStateProvider and then it threw an exception, please see below:

**Razor Page**
`@attribute [Authorize]`
...

**Code behind:**
-  ClaimsPrincipal anonymous = new(new ClaimsIdentity());
-  NotifyAuthenticationStateChanged(Task.FromResult(new AuthenticationState(anonymous)));
-  return Task.CompletedTask;

**Exception:**
```
blazor.webassembly.js:1 crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Collection was modified; enumeration operation may not execute.
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.Collections.Generic.List`1.Enumerator[[Microsoft.AspNetCore.Components.Rendering.ComponentState, Microsoft.AspNetCore.Components, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].MoveNextRare()
   at System.Collections.Generic.List`1.Enumerator[[Microsoft.AspNetCore.Components.Rendering.ComponentState, Microsoft.AspNetCore.Components, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].MoveNext()
   at Microsoft.AspNetCore.Components.CascadingValueSource`1.<>c__DisplayClass10_0[[System.Threading.Tasks.Task`1[[Microsoft.AspNetCore.Components.Authorization.AuthenticationState, Microsoft.AspNetCore.Components.Authorization, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<NotifyChangedAsync>b__0()
   at Microsoft.AspNetCore.Components.WebAssembly.Rendering.NullDispatcher.InvokeAsync(Action workItem)
   at Microsoft.AspNetCore.Components.CascadingValueSource`1[[System.Threading.Tasks.Task`1[[Microsoft.AspNetCore.Components.Authorization.AuthenticationState, Microsoft.AspNetCore.Components.Authorization, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].NotifyChangedAsync()
   at Microsoft.AspNetCore.Components.CascadingValueSource`1[[System.Threading.Tasks.Task`1[[Microsoft.AspNetCore.Components.Authorization.AuthenticationState, Microsoft.AspNetCore.Components.Authorization, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].NotifyChangedAsync(Task`1 newValue)
   at Microsoft.Extensions.DependencyInjection.CascadingAuthenticationStateServiceCollectionExtensions.AuthenticationStateCascadingValueSource.HandleAuthenticationStateChanged(Task`1 newAuthStateTask)
```

### .NET Version

NET 8

**Comment by uottoni** [comment]
Same here...
When I'm logginout the user i got this error too.
My custom AuthStateProvider:
`var annonymousUser = new ClaimsPrincipal(new ClaimsIdentity());
var authState = Task.FromResult(new AuthenticationState(annonymousUser));
NotifyAuthenticationStateChanged(authState);`
I got the error at last line: NotifyAuthenticationStateChanged(authState);
On my NET 7 project is working.

**Comment by patrickpham3339** [comment]
Not yet… the issue’s still there.On Jan 25, 2024, at 7:55 AM, Ali Jamal ***@***.***> wrote:﻿
same problem, i use blazorWebView with windows forms  and i have the same issue did you manage to fix it @patrickpham3339  ?

**Comment by bo-bac** [comment]
Pardone, why closed?
The same proble. Is it fixed?

**Comment by suntaurus** [comment]
I have the same issue. 
This is happening for me after I added builder.Services.AddCascadingAuthenticationState (.NET8) in the Program.cs.
If I comment this line and switch back to using <CascadingAuthenticationState> in the App.razor then there is no issue.
This happens when we logout and then immediately login back.

**Comment by R-Vulcan** [comment]
Can confirm, this seems to only happen while using AddCascadingAuthenticationState() instead of the CascadingAuthenticationState component.

The issue started occuring when I created a component in my layout that requested the cascading AuthState either from the CascadingParameter or the AuthorizeView.

**Comment by hyspdrt** [comment]
I have the same issue, on logging in. I too switched to the AddCascadingAuthenticationState() and had to revert back.

**Comment by PeterWhatsUpp** [comment]
This problem also occurs in Blazor Server mode. Invoke of NotifyAuthenticationStateChanged throws  System.Func`1 - "Collection was modified; enumeration operation may not execute" exception.
I've also changed AddCascadingAuthenticationState service to <CascadingAuthenticationState> to fix this issue.

**Comment by mabster** [comment]
Chiming in here - we've just seen the same issue. Reverting to <CascadingAuthenticationState> in Routes.razor instead of using the service has solved the problem for now. This issue should definitely not be closed.

**Comment by igotinfected** [comment]
Same issue here. Also fixed by reverting to `<CascadingAuthenticationState>`.

```stacktrace
at System.Collections.Generic.List`1.Enumerator[[Microsoft.AspNetCore.Components.Rendering.ComponentState, Microsoft.AspNetCore.Components, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].MoveNextRare()
   at System.Collections.Generic.List`1.Enumerator[[Microsoft.AspNetCore.Components.Rendering.ComponentState, Microsoft.AspNetCore.Components, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].MoveNext()
   at Microsoft.AspNetCore.Components.CascadingValueSource`1.<>c__DisplayClass10_0[[System.Threading.Tasks.Task`1[[Microsoft.AspNetCore.Components.Authorization.AuthenticationState, Microsoft.AspNetCore.Components.Authorization, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<NotifyChangedAsync>b__0()
   at Microsoft.AspNetCore.Components.WebAssembly.Rendering.NullDispatcher.InvokeAsync(Action workItem)
   at Microsoft.AspNetCore.Components.CascadingValueSource`1[[System.Threading.Tasks.Task`1[[Microsoft.AspNetCore.Components.Authorization.AuthenticationState, Microsoft.AspNetCore.Components.Authorization, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].NotifyChangedAsync()
   at Microsoft.AspNetCore.Components.CascadingValueSource`1[[System.Threading.Tasks.Task`1[[Microsoft.AspNetCore.Components.Authorization.AuthenticationState, Microsoft.AspNetCore.Components.Authorization, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].NotifyChangedAsync(Task`1 newValue)
   at Microsoft.Extensions.DependencyInjection.CascadingAuthenticationStateServiceCollectionExtensions.AuthenticationStateCascadingValueSource.HandleAuthenticationStateChanged(Task`1 newAuthStateTask)
   at Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider.NotifyAuthenticationStateChanged(Task`1 task)
   at Microsoft.AspNetCore.Components.WebAssembly.Authentication.RemoteAuthenticationService`3[[Microsoft.AspNetCore.Components.WebAssembly.Authentication.RemoteAuthenticationState, Microsoft.AspNetCore.Components.WebAssembly.Authentication, Version=8.0.6.0, Culture=neutral, PublicKeyToken=adb9793829ddae60],[<CompanyApplicationName>.Shared.Security.UserAccount, <CompanyApplicationName>.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[<CompanyApplicationName>.Oidc.Models.ExtendedOidcProviderOptions, <CompanyApplicationName>.Oidc, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].UpdateUser(Task`1 task)
   at Microsoft.AspNetCore.Components.WebAssembly.Authentication.RemoteAuthenticationService`3.<UpdateUserOnSuccess>d__31[[Microsoft.AspNetCore.Components.WebAssembly.Authentication.RemoteAuthenticationState, Microsoft.AspNetCore.Components.WebAssembly.Authentication, Version=8.0.6.0, Culture=neutral, PublicKeyToken=adb9793829ddae60],[<CompanyApplicationName>.Shared.Security.UserAccount, <CompanyApplicationName>.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[<CompanyApplicationName>.Oidc.Models.ExtendedOidcProviderOptions, <CompanyApplicationName>.Oidc, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()
```

**Comment by hannesSwart9** [comment]
It looks like NotifyAuthenticationStateChanged() causes the error if there is an <AuthorizeView> tag.
I use <AuthorizeView> in my navbar, and removing the navbar while testing gets rid of the error.
I see NotifyAuthenticationStateChanged accepts a task, but there is no option to invoke async.
So I'm suspecting the rendering breaks because it is not awaited and the <AuthorizeView> breaks because the authorized state changes while rendering.

**Comment by flier268** [comment]
So, I need to downgrade to <CascadingAuthenticationState>?

**Comment by Leon99** [comment]
@javiercn is there a way to stop auto-closing this issue (until it's fixed)?
I don't quite understand this policy. Regardless of whether the original poster responds or not, you still want to fix this, and ensure that all other people who encountered it (and subscribed to this) are notified, right?

**Comment by MackinnonBuck** [comment]
@thepra, we'll backport this if we get feedback that the .NET 9 fix fully addresses the problem. We're still in the prerelease phase for .NET 9, but we want to be cautious about changes we include in .NET 8 patch releases because the impact of introducing a regression is higher there.

We're very keen on getting feedback on this fix so that we can backport it. When .NET 9 RC1 releases, it would help us if you (and others) could confirm that it fixes the problem in your scenario.

Thanks for understanding!

**Comment by thepra** [comment]
Sounds good, lets wait.

**Comment by hyspdrt** [comment]
Updated to 8.0.8 with the patched code, and if I nest an AuthorizeView, I still encounter the exception. If I revert back to the `<CascadingAuthenticationState>`, nesting works just fine.

But, if I don't nest and only have 1 AuthorizeView, it does work and does not throw the error. So, it seems it's partially resolved.

Also, I'm using WASM Standalone, so the Dispatcher is the NullDispatcher. And the AuthorizeView is used in my Header.razor component which is on the MainLayout.razor, but the nested AuthorizeView is only used in the Authorized fragment of the parent AuthorizeView, so it only throws upon successful login.

Hope this feedback is helpful.

**Comment by janhjordie** [comment]
Same issue in dotnet 8.0.10. Just wasted several hours on figuring how to fix it

**Comment by mbrannen** [comment]
Ran into this issue myself.  Can confirm that updating to 9 resolved it.

**Comment by Poffertjie** [comment]
This is not just related to WASM but Server as well. How is this closed? Problem remains in .net 8 and forcing us to upgrade to .net 9 is just wrong.

**Comment by Conman-123** [comment]
A janky temporary solution for me was to add a handler for the `OnNavigateAsync` parameter of my `Router` in `App.razor`:
```
<Router AppAssembly="@typeof(App).Assembly" OnNavigateAsync="HandleNavigateStarted">

...

private void HandleNavigateStarted(NavigationContext context)
{
    StateHasChanged(); // This part isn't important, just having *something* here seems to fix it
}
```

**Comment by thepra** [comment]
> [@thepra](https://github.com/thepra), we'll backport this if we get feedback that the .NET 9 fix fully addresses the problem. We're still in the prerelease phase for .NET 9, but we want to be cautious about changes we include in .NET 8 patch releases because the impact of introducing a regression is higher there.
> 
> We're very keen on getting feedback on this fix so that we can backport it. When .NET 9 RC1 releases, it would help us if you (and others) could confirm that it fixes the problem in your scenario.
> 
> Thanks for understanding!

Is there any news?

**Comment by MackinnonBuck** [comment]
@thepra, the .NET 8 fix will be included in an upcoming servicing release (most likely in March).

**Comment by VJayMeyer** [comment]
Any updates on the lead time? I can see https://learn.microsoft.com/en-us/aspnet/core/blazor/security/?view=aspnetcore-8.0&tabs=visual-studio#troubleshoot-errors saying not to use <CascadingAuthenticationState> to rap the <router> but it's still the only solution.

**Comment by DanielCordell** [comment]
> Any updates on the lead time? I can see https://learn.microsoft.com/en-us/aspnet/core/blazor/security/?view=aspnetcore-8.0&tabs=visual-studio#troubleshoot-errors saying not to use to rap the but it's still the only solution.

@VJayMeyer being hopeful, looking at the MR to backport this change, https://github.com/dotnet/aspnetcore/pull/57288
it was merged last month and has the milestone `8.0.14` on it, which appears to have just come out?
